### PR TITLE
Fix FontTemplateDescriptor under FreeType

### PR DESCRIPTION
**Pull request by jonleighton**
Issue #17321. Under Linux, using "font-family: sans-serif" previously
caused Servo to select the "UltraLight" face (of DejaVu Sans). There were
two reasons for this:

1. Font weight was only retrieved from the OS/2 table for bold faces.
   This neglected to retrieve the weight information for "lighter than
   normal" weight faces. This meant that the UltraLight face appeared as
   normal weight, and was selected.

2. Retrieval of font stretch information from the OS/2 table was not
   implemented at all.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/19928)

### CODE REVIEW

**Comment by jonleighton** [comment]
I would like to implement tests for this change, but need some guidance. I asked about this in IRC but got no response:

> Hi! I was just looking at a font-related issue and decided to run tests/wpt/web-platform-tests/css/css-fonts on my machine. I got a bunch of failures, and then I realised that these tests are not run on the CI (it's configured in tests/wpt/include.ini). Why is that? What would be the best way to ensure a font-related change is properly tested?

I see two options (or both):

1. Enable the css-fonts WPT tests
2. Write unit tests

Advice please!

**Comment by jdm** [comment]
Go ahead and enable the tests. They were added when we weren't paying attention; there's no reason I know of not to run them in Servo.

**Comment by jonleighton** [comment]
I've been looking into how to test this and have hit several roadblocks after [enabling the css-fonts tests](https://github.com/servo/servo/pull/19930).

The change to how the font weight is detected is captured by [css/css-fonts/font-weight-normal-001.xht](https://github.com/w3c/web-platform-tests/blob/master/css/css-fonts/font-weight-normal-001.xht). (I haven't found an existing test which captures the font stretch fix.) However, it doesn't easily show this bug fix for two reasons:

1. That test is marked as a manual test in WPT, even though it's actually a reftest. Therefore it won't run on the CI. See https://github.com/w3c/web-platform-tests/issues/9374.
2. That test fails still fails after my fix due to presumably other bugs in how we are implementing the [font matching algorithm](https://drafts.csswg.org/css-fonts/#font-matching-algorithm). The rendering of the test changes, but it still doesn't pass (per my manual testing with and without this patch).

Therefore I propose to write a unit test to encapsulate this change.

**Comment by jdm** [comment]
A unit test sounds sensible, in that case.

**Comment by jonleighton** [comment]
OK, I've added a test. I tried to use the existing test fonts in `tests/wpt/web-platform-tests/css/fonts/CSSTest/`, but there were no condensed/expanded versions to test the font-stretch detection. So I added the DejaVu fonts. The addition of the `descriptor()` method yielded a nice cleanup for the existing methods which load the descriptor.

File: components/gfx/font_context.rs [hunk 1]
```diff
@@ -9,7 +9,7 @@ use font_cache_thread::FontCacheThread;
 use font_template::FontTemplateDescriptor;
 use malloc_size_of::{MallocSizeOf, MallocSizeOfOps};
 use platform::font::FontHandle;
-use platform::font_context::FontContextHandle;
+pub use platform::font_context::FontContextHandle;
```
Comment (jonleighton) [inline components/gfx/font_context.rs]: I had to do this for the test, which doesn't seem ideal but I struggled to find an alternative. In the existing code, a `FontContext` instantiates a `FontContextHandle`, but instantiating a `FontContext` requires lots of dependencies (some of them external to this crate too).

**Comment by jonleighton** [comment]
The test I wrote failed on Windows but it's fixed now.

**Comment by jonleighton** [comment]
@emilio has reviewed and says he's happy with the code, but he's unsure about whether there are licensing issues caused by the inclusion of the DejaVu fonts for testing: https://mozilla.logbot.info/servo/20180206#c14250864-c14250932

**Comment by metajack** [comment]
The license is fine.

**Comment by jonleighton** [comment]
Hrm, I'll try to figure out why that test fails on mac (though if anyone has easy access to a mac and can help that would be welcome...)

**Comment by jdm** [comment]
I wonder if the rounding in https://github.com/jonleighton/servo/blob/ecf441dffc49e282623613d96841ce527f518978/components/gfx/platform/macos/font.rs#L210-L218 is influencing that result.

**Comment by jonleighton** [comment]
Seems likely, in which case the fix should be straightforward. I'll take a proper look in a bit.

**Comment by jdm** [comment]
I've stepped through it, and the values coming out of the normalization process are very close to 6.0, so the rounding is not interfering.

**Comment by jdm** [comment]
Given the list of font weights at https://git.merproject.org/mer-core/qtbase/commit/9ba296cc4cefaeb9d6c5abc2e0c0b272f2288733#1b84d1913347bd20dd0a134247f8cd012a646261_44_52 and the fact that they are not evenly distributed like the weights from 100-900, I can see why the conversion implemented in `boldness` doesn't not map to the expected weight.

**Comment by jdm** [comment]
We may just want to make the macOS implementation rely on the OS/2 table as well, rather than trying to map between what CoreText reports.

**Comment by jonleighton** [comment]
Thanks for the research! I'll see whether I can make some non-platform-specific code to read the OS/2 table.

**Comment by jonleighton** [comment]
It seems that CoreText can also provide a different kind of weight value which is [an integer from 0 to 14](https://developer.apple.com/documentation/appkit/nsfontmanager/1462351-weight). Gecko maps these numbers to CSS numbers [via a static table](https://github.com/mozilla/gecko-dev/blob/master/gfx/thebes/gfxMacPlatformFontList.mm#L94-L110), so that's another option I guess.

**Comment by jonleighton** [comment]
Just read through [this enormous bug](https://bugzilla.mozilla.org/show_bug.cgi?id=931426) about Gecko's approach on OS X. Basically it appears that some of the fonts shipped with OS X contain a large number of faces, and CoreText reports some funky weight values for those faces (in some cases, lighter faces were reported with a higher weight).

They iterated on lots of approaches to fix it, including reading the OS/2 table directly. However, apparently some of the fonts shipped in OS X contained the *same* OS/2 weight value for different faces, among other issues.

In the end, they implemented a pref which allows certain known problematic fonts to have their weight value overridden. The code for that is [here](https://github.com/mozilla/gecko-dev/blob/106842f84295b881f87c8cb54b17390450b94a0f/gfx/thebes/gfxMacPlatformFontList.mm#L829-L841).

This makes me think that we probably should not rely on the OS/2 table on Mac.

**Comment by jonleighton** [comment]
I think I'm going to struggle to come up with a solution for Mac given that I don't have access to one. Is anybody else willing to help with this? Alternatively, could we merge this fix with the test running on Windows & Linux only, and then make a note that the Mac handling needs refinement?

**Comment by metajack** [comment]
@jonleighton Would ssh access to one of our builders be sufficient? I'm sure we can get you hooked up with that for testing if it would help you make progress.

Alternatively, maybe @pcwalton or @mbrubeck can help who know quite a bit about fonts.

**Comment by pcwalton** [comment]
I'm fine with just landing this for Windows and Linux and doing Mac in a followup. In fact, given how complex this will likely be on Mac, that's probably preferable, just for reasons of keeping the size of this patch under control.

**Comment by jonleighton** [comment]
Ok, I agree with @pcwalton that we should leave the Mac work for a follow-up, so I have updated my patch to compile out the test under Mac.

That said, I'm up for doing the follow-up Mac work, so access to a builder would be great @metajack. (I'd like to do more work on fonts too, so this would be really helpful in general.) You can email me at j at jonathanleighton dot com to arrange. Thanks!

**Comment by jdm** [comment]
@jonleighton You'll want to submit a PR adding your public key to https://github.com/servo/saltfs/tree/master/admin/files/ssh in that case.

**Comment by jonleighton** [comment]
@jdm: 👍 

Looks like the latest Travis CI failure is unrelated - https://github.com/servo/servo/issues/14774

Could somebody r+ this again please?

**Comment by jdm** [comment]
@bors-servo retry
* freed up space on servo-linux6

**Comment by jdm** [comment]
@bors-servo retry
* #13509

**Pull request merged**